### PR TITLE
High Resolution Metrics Support

### PR DESCRIPTION
**Pull request by meshwa19**
*Description of changes:*

Added support for High Resolution Metrics. Customers can specify an optional property - storageResolution of the metrics in putMetric() call. If resolution is not specified, metrics will be emitted with standard resolution.
Added UTs and modified integration tests for the changes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/logger/MetricsContext.ts [hunk 1]
```diff
@@ -180,14 +182,19 @@ export class MetricsContext {
...
+    storageResolution?: StorageResolution | number,
+  ): void {
+    validateMetric(key, value, unit, storageResolution, this.metricNameAndResolutionMap);
 
     const currentMetric = this.metrics.get(key);
     if (currentMetric) {
       currentMetric.addValue(value);
     } else {
-      this.metrics.set(key, new MetricValues(value, unit));
+      this.metrics.set(key, new MetricValues(value, unit, storageResolution));
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply src/logger/MetricsContext.ts:197]: This is being done [here](https://github.com/awslabs/aws-embedded-metrics-node/pull/142/files#diff-60dededc3634fc2b328d5990d252b89a2ca90599d988b79e0b65bd607ce8ab21R138).

File: src/logger/__tests__/MetricsContext.test.ts [hunk 2]
```diff
@@ -203,6 +223,23 @@ test('putMetric uses None unit if not provided', () => {
   expect(metricDatum.unit).toBe(expectedUnit);
 });
 
+test('putMetric uses Standard storageResolution if not provided', () => {
+  // arrange
+  const context = MetricsContext.empty();
+  const expectedKey = faker.random.word();
+  const expectedValue = faker.datatype.number();
+  const expectedStorageResolution = 60; //TODO_M remove this after metric validation
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply src/logger/__tests__/MetricsContext.test.ts:231]: Removed it. Thanks!

File: examples/lambda/src/index.js [hunk 3]
```diff
@@ -1,9 +1,10 @@
-const { metricScope } = require('aws-embedded-metrics');
+const { metricScope , StorageResolution} = require('aws-embedded-metrics');
 
 const aggregator = metricScope(metrics => async event => {
   console.log('received message');
   metrics.putDimensions({ Service: 'Aggregator' });
   metrics.putMetric('ProcessingLatency', 100, 'Milliseconds');
+  metrics.putMetric('CPU Utilization', 87, 'Percent', 1);
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply examples/lambda/src/index.js]: Updated, thanks!

File: src/logger/MetricsContext.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (meshwa19) [reply src/logger/MetricsContext.ts:197]: Move it here from Validator class

**Review by rayabagi — APPROVED** [review]
Finished with the review.

**Comment by markkuhn** [comment]
Should we add information about `StorageResolution` in the `README`? Maybe the 2 types of resolution and a link the relevant wiki?

**Comment by markkuhn** [comment]
nit: run prettier formatting on all your changes

File: src/logger/__tests__/MetricsContext.test.ts [hunk 4]
```diff
@@ -353,6 +390,20 @@ test.each([
...
+test('put metric with same key and different resolution in single flush throws error' ,() => {
+  //arrange
+  const context = MetricsContext.empty();
+  const expectedKey = 'MetricName';
+  const expectedValue = faker.datatype.number();
+  const expectedUnit = 'None';
+
+  // act 
+ expect(() => {
+    context.putMetric(expectedKey, expectedValue,expectedUnit,StorageResolution.High);
```
Comment (markkuhn) [inline src/logger/__tests__/MetricsContext.test.ts:402]: `context.putMetric(expectedKey, expectedValue,expectedUnit,StorageResolution.High)` should not be expected to throw an error.

File: src/serializers/LogSerializer.ts [hunk 5]
```diff
@@ -108,9 +109,16 @@ export class LogSerializer implements ISerializer {
...
         // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
         currentBody[metricProgress.name] = metricValue;
+        const metricBody: {[key: string]: any} = {
+          Name: metricProgress.name,
+          Unit: metric.unit,
+        }
+        if (metric.storageResolution === StorageResolution.High)
+        {
+          metricBody.StorageResolution = StorageResolution.High;
+        }
```
Comment (markkuhn) [inline src/serializers/LogSerializer.ts:119]:
```suggestion
        const metricBody: {[key: string]: any} = {
          Name: metricProgress.name,
          Unit: metric.unit,
          ...(metric.storageResolution == StorageResolution.High ? { StorageResolution: StorageResolution.High } : {}),
        }
```

File: src/utils/Validator.ts [hunk 6]
```diff
@@ -116,6 +125,23 @@ const validateMetric = (key: string, value: number, unit?: Unit | string): void
...
+  if (
+    storageResolution !== undefined &&
+    !Object.values(StorageResolution)
+      .map((s) => s)
+      .includes(storageResolution)
+  ) {
+    throw new InvalidMetricError(`Metric resolution ${storageResolution} is not valid`);
+  }
+
+  if (metricNameAndResolutionMap && metricNameAndResolutionMap.has(key)) {
```
Comment (markkuhn) [inline src/utils/Validator.ts:138]: Why are we checking if `metricNameAndResolutionMap` is null/undefined?

File: test/integ/agent/end-to-end.integ.ts [hunk 7]
```diff
@@ -57,7 +58,7 @@ test(
 
     const doWork = metricScope(metrics => () => {
       metrics.putDimensions(dimensions);
-      metrics.putMetric(metricName, 100, 'Milliseconds');
+      metrics.putMetric(metricName, 100, 'Milliseconds', StorageResolution.High);
```
Comment (markkuhn) [inline test/integ/agent/end-to-end.integ.ts:61]: Can we make this one `Standard` since the one on line 37 is `High`.

Comment (meshwa19) [reply test/integ/agent/end-to-end.integ.ts]: Done.

File: src/serializers/LogSerializer.ts (hunk 5 above)
Comment (meshwa19) [reply src/serializers/LogSerializer.ts]: Thanks for the suggestion!! Updated.

File: src/logger/__tests__/MetricsContext.test.ts (hunk 4 above)
Comment (meshwa19) [reply src/logger/__tests__/MetricsContext.test.ts]: Updated!

File: src/utils/Validator.ts [hunk 8]
```diff
@@ -116,6 +125,23 @@ const validateMetric = (key: string, value: number, unit?: Unit | string): void
   ) {
     throw new InvalidMetricError(`Metric unit ${unit} is not valid`);
   }
+
+  if (
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply src/utils/Validator.ts:129]: Added!

**Comment by meshwa19** [comment]
> nit: run prettier formatting on all your changes

Done!

File: src/logger/MetricsContext.ts [hunk 9]
```diff
@@ -180,15 +182,24 @@ export class MetricsContext {
...
     }
+    if (this.metricNameAndResolutionMap && !this.metricNameAndResolutionMap.has(key)) {
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply src/logger/MetricsContext.ts]: Not needed. Removed it. Thanks!

File: src/utils/Validator.ts (hunk 6 above)
Comment (meshwa19) [reply src/utils/Validator.ts]: Yes, not needed. Removed it!

File: src/logger/__tests__/MetricsContext.test.ts [hunk 10]
```diff
@@ -314,43 +351,67 @@ test.each([
...
+  [faker.random.word(), parseFloat('not a number'), undefined, undefined],
+  [faker.random.word(), Infinity, faker.helpers.arrayElement(Object.values(Unit)), undefined],
+  [faker.random.word(), -Infinity, faker.helpers.arrayElement(Object.values(Unit)), undefined],
+  [faker.random.word(), faker.datatype.number(), 'Fahrenheit', undefined],
+  [faker.random.word(), 4, '', undefined],
+  [faker.random.word(), NaN, faker.helpers.arrayElement(Object.values(Unit)), undefined],
+  [faker.random.words(3), faker.datatype.number(), Unit.Seconds, 45],
+  [faker.random.words(3), faker.datatype.number(), Unit.Seconds, 0],
+])(
+  'putMetric with name: %s, value: %d and unit: %s throws error',
```
Comment (markkuhn) [inline src/logger/__tests__/MetricsContext.test.ts:369]:
```suggestion
  'putMetric with name: %s, value: %d, unit: %s and resolution: %d throws error',
```

File: src/logger/__tests__/MetricsContext.test.ts [hunk 11]
```diff
@@ -314,43 +351,67 @@ test.each([
...
+  [faker.random.word(), faker.datatype.number({ min: -1e3, max: -1 }), undefined, undefined],
+  [faker.random.word(), faker.datatype.number(), faker.helpers.arrayElement(Object.values(Unit)), undefined],
+  [faker.random.words(2), faker.datatype.number(), undefined, undefined],
+  [faker.random.words(3), faker.datatype.number(), Unit.Seconds, undefined],
+  ['Max_Value', Number.MAX_SAFE_INTEGER, Unit.Milliseconds, undefined],
+  ['-Max_Value', -Number.MAX_SAFE_INTEGER, 'Bytes/Second', undefined],
+  ['-Max_Value', faker.datatype.number(), 'Bytes/Second', 1],
+  ['-Max_Value', faker.datatype.number(), 'Bytes/Second', 60],
+])(
+  'putMetric with name: %s, value: %d and unit: %s does not throw error',
```
Comment (markkuhn) [inline src/logger/__tests__/MetricsContext.test.ts:391]:
```suggestion
  'putMetric with name: %s, value: %d, unit: %s and resolution: %d does not throw error',
```

File: src/logger/__tests__/MetricsContext.test.ts [hunk 12]
```diff
@@ -385,8 +385,8 @@ test.each([
   [faker.random.words(3), faker.datatype.number(), Unit.Seconds, undefined],
   ['Max_Value', Number.MAX_SAFE_INTEGER, Unit.Milliseconds, undefined],
   ['-Max_Value', -Number.MAX_SAFE_INTEGER, 'Bytes/Second', undefined],
-  ['-Max_Value', -Number.MAX_SAFE_INTEGER, 'Bytes/Second', 1],
-  ['-Max_Value', -Number.MAX_SAFE_INTEGER, 'Bytes/Second', 60],
+  ['-Max_Value', faker.datatype.number(), 'Bytes/Second', 1],
+  ['-Max_Value', faker.datatype.number(), 'Bytes/Second', 60],
```
Comment (markkuhn) [inline src/logger/__tests__/MetricsContext.test.ts:389]: Why did you change from `-Number.MAX_SAFE_INTEGER` to `faker.datatype.number()`?

Comment (meshwa19) [reply src/logger/__tests__/MetricsContext.test.ts:389]: The last two lines of the test verify storage resolution and they are not to verify edge cases for the unit parameter (they are getting verified in the lines above). So I rectified my mistake by updating the unit value to some random value instead of an edge case.

File: src/logger/__tests__/MetricsContext.test.ts (hunk 10 above)
Comment (meshwa19) [reply src/logger/__tests__/MetricsContext.test.ts]: Done, thanks!

File: src/logger/__tests__/MetricsContext.test.ts (hunk 11 above)
Comment (meshwa19) [reply src/logger/__tests__/MetricsContext.test.ts]: Done, thanks!

**Review by markkuhn — APPROVED** [review]
🚀
